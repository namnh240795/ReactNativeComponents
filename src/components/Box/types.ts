import {FlexWrap} from './../../atoms/theme/constants/common';
import {
  Align,
  AlignSelf,
  FlexDirection,
  Justify,
  Spacing,
} from '@src/atoms/theme/constants/common';
import {ThemeColor} from '@src/atoms/theme/constants/types';
import {ReactNode} from 'react';
import {TouchableOpacityProps, ViewProps} from 'react-native';

export interface BoxProps extends ViewProps, TouchableOpacityProps {
  background?: ThemeColor;
  flexDirection?: FlexDirection;
  justify?: Justify;
  align?: Align;
  alignSelf?: AlignSelf;
  flex?: number;
  square?: number;
  circle?: number | string;
  width?: string | number | undefined;
  height?: string | number | undefined;
  minWidth?: string | number | undefined;
  minHeight?: string | number | undefined;
  margin?: Spacing;
  padding?: Spacing;
  pressable?: boolean;
  borderRadius?: Spacing;
  children?: ReactNode;
  flexWrap?: FlexWrap;
  gap?: number;
  columnGap?: number;
  rowGap?: number;
}
