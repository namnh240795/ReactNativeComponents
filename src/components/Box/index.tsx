import { stylesUtils } from '@src/atoms/theme/constants/common';
import React, { useMemo } from 'react';
import { StyleProp, StyleSheet } from 'react-native';
import { TouchableOpacity, View, ViewStyle } from 'react-native';
import { BoxProps } from './types';

const Box: React.FC<BoxProps> = ({
  pressable,
  flexDirection,
  justify,
  align,
  alignSelf,
  flex,
  background,
  square,
  circle,
  height,
  width,
  minWidth,
  minHeight,
  margin,
  padding,
  borderRadius,
  children,
  flexWrap,
  onPress,
  style,
  ...props
}) => {
  const combinedStyle: StyleProp<ViewStyle> = useMemo(() => {
    let marginValues, paddingValues;
    let borderRadiusValues;
    let flexValue,
      widthValue,
      heightValue,
      squareValue,
      circleValue,
      minWidthVal,
      minHeightVal,
      flexWrapVal;
    if (flexWrap) {
      flexWrapVal = stylesUtils.flexWrap(flexWrap);
    }
    if (flex) {
      flexValue = stylesUtils.flex(flex);
    }
    if (width) {
      widthValue = stylesUtils.width(width);
    }
    if (height) {
      heightValue = stylesUtils.height(height);
    }
    if (square) {
      squareValue = stylesUtils.square(square);
    }
    if (circle) {
      circleValue = stylesUtils.circle(circle);
    }
    if (minWidth) {
      minWidthVal = stylesUtils.minWidth(minWidth);
    }
    if (minHeight) {
      minHeightVal = stylesUtils.minHeight(minHeight);
    }
    if (margin) {
      marginValues = stylesUtils.margin(stylesUtils.normalizeOptions(margin));
    }
    if (padding) {
      paddingValues = stylesUtils.padding(
        stylesUtils.normalizeOptions(padding),
      );
    }
    if (borderRadius) {
      borderRadiusValues = stylesUtils.borderRadius(
        stylesUtils.normalizeOptions(borderRadius),
      );
    }

    return StyleSheet.flatten([
      marginValues,
      paddingValues,
      borderRadiusValues,
      background && stylesUtils.background(background),
      flexDirection && stylesUtils.flexDirection(flexDirection),
      justify && stylesUtils.justify(justify),
      align && stylesUtils.align(align),
      alignSelf && stylesUtils.alignSelf(alignSelf),
      flexValue,
      widthValue,
      heightValue,
      squareValue,
      circleValue,
      minHeightVal,
      minWidthVal,
      flexWrapVal,
      style,
    ]);
  }, [
    align,
    alignSelf,
    background,
    flexDirection,
    justify,
    margin,
    padding,
    borderRadius,
    flex,
    width,
    height,
    minWidth,
    minHeight,
    square,
    circle,
    style,
    flexWrap,
  ]);

  if (pressable) {
    return (
      <TouchableOpacity
        activeOpacity={0.7}
        style={combinedStyle}
        onPress={onPress}
        {...props}
      >
        {children}
      </TouchableOpacity>
    );
  }

  return (
    <View style={combinedStyle} {...props}>
      {children}
    </View>
  );
};

export default Box;
