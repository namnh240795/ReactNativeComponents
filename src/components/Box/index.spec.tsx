import React from 'react';
import renderer from 'react-test-renderer';
import Box from '.';

test('Render empty box', () => {
  const tree = renderer.create(<Box />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render GOLBAT_120 box', () => {
  const tree = renderer.create(<Box background={'GOLBAT_120'} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render custom color square 40 box', () => {
  const tree = renderer
    .create(<Box background="#ffffff" square={40} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render DITTO_140 circle 40 box', () => {
  const tree = renderer
    .create(<Box background="DITTO_140" circle={40} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render padding all box', () => {
  const tree = renderer.create(<Box padding={8} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render padding hv box', () => {
  const tree = renderer.create(<Box padding={[0, 2]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render padding width array box', () => {
  const tree = renderer.create(<Box padding={[0, 2, 0, 4]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin undefined box', () => {
  const tree = renderer.create(<Box margin={undefined} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin all box', () => {
  const tree = renderer.create(<Box margin={8} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin hv box', () => {
  const tree = renderer.create(<Box margin={[0, 2]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin with number array box', () => {
  const tree = renderer.create(<Box margin={[0, 2, 0, 4]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render border all box', () => {
  const tree = renderer.create(<Box borderRadius={8} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render flex 1 box', () => {
  const tree = renderer.create(<Box flex={1} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render flexWrap box', () => {
  const tree = renderer.create(<Box flexWrap="wrap" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render width 100 box', () => {
  const tree = renderer.create(<Box width={100} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render height 100 box', () => {
  const tree = renderer.create(<Box height={100} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render min width 100 box', () => {
  const tree = renderer.create(<Box minWidth={100} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render min height 100 box', () => {
  const tree = renderer.create(<Box minHeight={100} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render pressable box', () => {
  const tree = renderer.create(<Box pressable />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render align box', () => {
  const tree = renderer
    .create(
      <Box
        align="center"
        alignSelf="center"
        justify="center"
        flexDirection="row"
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
