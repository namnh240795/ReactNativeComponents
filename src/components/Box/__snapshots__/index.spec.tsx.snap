// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render DITTO_140 circle 40 box 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#565CE3",
      "borderRadius": 20,
      "height": 40,
      "width": 40,
    }
  }
/>
`;

exports[`Render GOLBAT_120 box 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#343A40",
    }
  }
/>
`;

exports[`Render align box 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
/>
`;

exports[`Render border all box 1`] = `
<View
  style={
    Object {
      "borderBottomLeftRadius": 8,
      "borderBottomRightRadius": 8,
      "borderTopLeftRadius": 8,
      "borderTopRightRadius": 8,
    }
  }
/>
`;

exports[`Render custom color square 40 box 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": 40,
      "width": 40,
    }
  }
/>
`;

exports[`Render empty box 1`] = `
<View
  style={Object {}}
/>
`;

exports[`Render flex 1 box 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`Render flexWrap box 1`] = `
<View
  style={
    Object {
      "flexWrap": "wrap",
    }
  }
/>
`;

exports[`Render height 100 box 1`] = `
<View
  style={
    Object {
      "height": 100,
    }
  }
/>
`;

exports[`Render margin all box 1`] = `
<View
  style={
    Object {
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
      "marginTop": 8,
    }
  }
/>
`;

exports[`Render margin hv box 1`] = `
<View
  style={
    Object {
      "marginBottom": 0,
      "marginLeft": 2,
      "marginRight": 2,
      "marginTop": 0,
    }
  }
/>
`;

exports[`Render margin undefined box 1`] = `
<View
  style={Object {}}
/>
`;

exports[`Render margin with number array box 1`] = `
<View
  style={
    Object {
      "marginBottom": 0,
      "marginLeft": 4,
      "marginRight": 2,
      "marginTop": 0,
    }
  }
/>
`;

exports[`Render min height 100 box 1`] = `
<View
  style={
    Object {
      "minHeight": 100,
    }
  }
/>
`;

exports[`Render min width 100 box 1`] = `
<View
  style={
    Object {
      "minWidth": 100,
    }
  }
/>
`;

exports[`Render padding all box 1`] = `
<View
  style={
    Object {
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    }
  }
/>
`;

exports[`Render padding hv box 1`] = `
<View
  style={
    Object {
      "paddingBottom": 0,
      "paddingLeft": 2,
      "paddingRight": 2,
      "paddingTop": 0,
    }
  }
/>
`;

exports[`Render padding width array box 1`] = `
<View
  style={
    Object {
      "paddingBottom": 0,
      "paddingLeft": 4,
      "paddingRight": 2,
      "paddingTop": 0,
    }
  }
/>
`;

exports[`Render pressable box 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
/>
`;

exports[`Render width 100 box 1`] = `
<View
  style={
    Object {
      "width": 100,
    }
  }
/>
`;
